### PR TITLE
DCS failsafe mode behave tests

### PR DESCRIPTION
**Pull request by hughcapet**

### CODE REVIEW

File: features/dcs_failsafe_mode.feature [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    And I sleep for 2 seconds
+    Then postgres1 role is the replica after 12 seconds
+
+  Scenario: check leader and replica are functioning while DCS is down
+    Given I start postgres0
+    Then "members/postgres0" key in DCS has state=running after 10 seconds
+    And "members/postgres1" key in DCS has state=running after 2 seconds
+    And logical slot dcs_slot_0 is in sync between postgres0 and postgres1 after 10 seconds
+    When I issue a GET request to http://127.0.0.1:8009/failsafe
+    Then I receive a response code 200
```
Comment (CyberDem0n) [inline features/dcs_failsafe_mode.feature:41]:
```suggestion
    Then I receive a response code 200
    And I receive a response postgres0 http://127.0.0.1:8008/patroni
    And I receive a response postgres1 http://127.0.0.1:8009/patroni
```

File: features/dcs_failsafe_mode.feature [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    Given DCS is down
+    And I sleep for 12 seconds
+    Then postgres0 role is the primary after 2 seconds
+    Then postgres1 role is the replica after 2 seconds
+    When I create table on postgres0
+    Then table mytest is present on postgres1 after 10 seconds
```
Comment (CyberDem0n) [inline features/dcs_failsafe_mode.feature]:
```suggestion
    And replication works from postgres0 to postgres1 after 10 seconds
```

File: features/dcs_failsafe_mode.feature [hunk 3]
```diff
@@ -0,0 +1,46 @@
+Feature:
```
Comment (CyberDem0n) [inline features/dcs_failsafe_mode.feature]:
```suggestion
Feature: dcs failsafe mode
```

It controls the `output` directory for patroni configs and various logs.

File: features/steps/dcs_failsafe_mode.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+@step('DCS is down')
+def stop_dcs(context):
+    if context.dcs_ctl._handle:
+        context.dcs_ctl._handle.send_signal(signal.SIGSTOP)
+
+
+@step('DCS is up')
+def start_dcs(context):
+    if context.dcs_ctl._handle:
+        context.dcs_ctl._handle.send_signal(signal.SIGCONT)
```
Comment (CyberDem0n) [inline features/steps/dcs_failsafe_mode.py]: I would suggest moving the actual implementation to the `AbstractDcsController`. Methods could be called `start_outage()` and `stop_outage()`.
It makes it possible to override methods for DCS that are not started by behave.

In addition to that please add `context.dcs_ctl.stop_outage()` to the `after_feature()` function. This is necessary for graceful shutdown of behave.

File: features/dcs_failsafe_mode.feature [hunk 5]
```diff
@@ -29,16 +30,22 @@ Feature: dcs failsafe mode
...
 
-  Scenario: check leader and replica are functioning while DCS is down
+  @dcs-failsafe
+  Scenario: check leader and replica are both in /failsafe key after leader is back
     Given I start postgres0
     Then "members/postgres0" key in DCS has state=running after 10 seconds
     And "members/postgres1" key in DCS has state=running after 2 seconds
-    And logical slot dcs_slot_0 is in sync between postgres0 and postgres1 after 10 seconds
-    When I issue a GET request to http://127.0.0.1:8009/failsafe
+    When I sleep for 2 seconds
```
Comment (hughcapet) [inline features/dcs_failsafe_mode.feature]: Most likely it should be more than 2 second here...

File: features/dcs_failsafe_mode.feature [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    And postgres1 role is the replica after 2 seconds
+    And replication works from postgres0 to postgres1 after 10 seconds
+    And I get all changes from logical slot dcs_slot_0 on postgres0
+    And logical slot dcs_slot_0 is in sync between postgres0 and postgres1 after 10 seconds
+
+ @dcs-failsafe
+  Scenario: check three-node cluster is functioning while DCS is down
+    Given DCS is up
+    When I do a backup of postgres0
+    And I start postgres2 in a cluster batman from backup with no_master
```
Comment (CyberDem0n) [inline features/dcs_failsafe_mode.feature]: It could be just:
```suggestion
    And I start postgres2
```

Comment (hughcapet) [reply features/dcs_failsafe_mode.feature]: sure...

File: features/dcs_failsafe_mode.feature [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    And postgres1 role is the replica after 2 seconds
+    And replication works from postgres0 to postgres1 after 10 seconds
+    And I get all changes from logical slot dcs_slot_0 on postgres0
+    And logical slot dcs_slot_0 is in sync between postgres0 and postgres1 after 10 seconds
+
+ @dcs-failsafe
+  Scenario: check three-node cluster is functioning while DCS is down
+    Given DCS is up
+    And I start postgres2
+    Then "members/postgres2" key in DCS has state=running after 10 seconds
```
Comment (hughcapet) [inline features/dcs_failsafe_mode.feature]: @CyberDem0n shouldn't we also check response for `GET http://127.0.0.1:8009/failsafe` here...?

Comment (CyberDem0n) [reply features/dcs_failsafe_mode.feature]: Do you expect it not to work despite working at https://github.com/CyberDem0n/patroni/pull/4/files#diff-e200798e6805f00fc721233822d3b238e7f9eb81187ec5febd68df6f79747d6cR39-R42?

Comment (hughcapet) [reply features/dcs_failsafe_mode.feature]: Can't remember the point I wrote this yesterday. Seems it wasn't that relevant anyway...

File: features/dcs_failsafe_mode.feature [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    And I receive a response postgres0 http://127.0.0.1:8008/patroni
+    When I issue a PATCH request to http://127.0.0.1:8008/config with {"postgresql": {"parameters": {"wal_level": "logical"}}}
+    Then I receive a response code 200
+    When I issue a PATCH request to http://127.0.0.1:8008/config with {"slots": {"dcs_slot_0": {"type": "logical", "database": "postgres", "plugin": "test_decoding"}}}
+    Then I receive a response code 200
+
+  @dcs-failsafe
+  Scenario: check one-node cluster is functioning while DCS is down
+    Given DCS is down
+    Then postgres0 role is the primary after 20 seconds
```
Comment (CyberDem0n) [inline features/dcs_failsafe_mode.feature]: I think here we need to have a sleep 12 before check.

Comment (hughcapet) [reply features/dcs_failsafe_mode.feature]:
```
    Given DCS is down
    And I sleep for 12 seconds
    Then postgres0 role is the primary after 2 seconds
```
like this?

Comment (CyberDem0n) [reply features/dcs_failsafe_mode.feature]: Then postgres0 role is the primary after 10 seconds

When executed on GH Actions we use multiplier 5: https://github.com/zalando/patroni/blob/master/features/environment.py#L878
2*5 == 10s, which still might be not enough for promote to finish on MacOS. With 10s we get the timeout 50s.

Comment (hughcapet) [reply features/dcs_failsafe_mode.feature]: Ah, now I get it, thanks

File: features/dcs_failsafe_mode.feature [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    When I do a backup of postgres0
+    And I shut down postgres0
+    And DCS is up
+    When I start postgres1 in a cluster batman from backup with no_master
+    And I sleep for 2 seconds
+    Then postgres1 role is the replica after 12 seconds
+
+  @dcs-failsafe
+  Scenario: check leader and replica are both in /failsafe key after leader is back
+    Given I start postgres0
```
Comment (CyberDem0n) [inline features/dcs_failsafe_mode.feature]: I would suggest the following change so we can at least test that the /failsafe key in DCS is properly updated when we can't simulate outage:
```suggestion
  Scenario: check leader and replica are both in /failsafe key after leader is back
    Given I start postgres0
    And I start postgres1
```

**Pull request closed without merging**